esx_service/utils/vm_listener.py

```diff
@@ -0,0 +1,205 @@
+    """
+    pc, error_msg = get_propertycollector()
+
+    if error_msg:
+        logging.warn("Could not start VM Listener: %s", error_msg)
+        return
+
+    # listen to changes
+    listen_vm_propertychange(pc)
+
+
+def create_vm_powerstate_filter(pc, from_node):
+    """
+    Create a filter spec to list to VM power state changes
+    """
+
+    filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+    objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=from_node,
+                                                       selectSet=vm_folder_traversal())
+    filterSpec.objectSet.append(objSpec)
+    # Add the property specs
+    propSpec = vmodl.query.PropertyCollector.PropertySpec(type=vim.VirtualMachine, all=False)
+    propSpec.pathSet.append(VM_POWERSTATE)
+    filterSpec.propSet.append(propSpec)
+    try:
+        pcFilter = pc.CreateFilter(filterSpec, True)
+        atexit.register(pcFilter.Destroy)
+        return None
+    except Exception as e:
+        err_msg = "Problem creating PropertyCollector filter: {}".format(str(e))
+        logging.error(err_msg)
+        return err_msg
+
+
+def listen_vm_propertychange(pc):
+    """
+    Waits for updates on powerstate of VMs. If powerstate is poweroff,
+    detach the dvs managed volumes attached to VM.
+    """
+    logging.info("VMChangeListener thread started")
+    version = ''
+    while True:
+        try:
+            result = pc.WaitForUpdates(version)
+            # process the updates result
+            for filterSet in result.filterSet:
+                for objectSet in filterSet.objectSet:
+                    if objectSet.kind != 'modify':
+                        continue
+                    for change in objectSet.changeSet:
+                        # if the event was powerOff for a VM, set the status of all
+                        # docker volumes attached to the VM to be detached
+                        if change.name != VM_POWERSTATE or change.val != POWERSTATE_POWEROFF:
+                            continue
+
+                        moref = getattr(objectSet, 'obj', None)
+                        # Do we need to alert the admin? how?
+                        if not moref:
+                            logging.error("Could not retrieve the VM managed object.")
+                            continue
+
+                        logging.info("VM poweroff change found for %s", moref.config.name)
+
+                        set_device_detached(moref)
+            version = result.version
+        # Capture hostd down exception
+        except http.client.RemoteDisconnected as e:
+            logging.error("VMChangeListener: Hostd connection error %s", str(e))
+            version = ''
+            # Need to get new SI instance, create a new property collector and property filter
+            # for it. Can't use the old one due to stale authentication error.
+            pc, error_msg = get_propertycollector()
```

Comment:
This is fine. Just curious why not exit thread and start a new thread start_vm_changelistener?